rosidl_runtime_c/include/rosidl_runtime_c/type_hash.h

```diff
@@ -0,0 +1,84 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSIDL_RUNTIME_C__TYPE_HASH_H_
+#define ROSIDL_RUNTIME_C__TYPE_HASH_H_
+
+#include <stdint.h>
+
+#include "rcutils/allocator.h"
+#include "rcutils/sha256.h"
+
+#include "rosidl_runtime_c/visibility_control.h"
+
+#define ROSIDL_TYPE_HASH_VERSION_UNSET 0
+#define ROSIDL_TYPE_HASH_SIZE RCUTILS_SHA256_BLOCK_SIZE
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+/// A ROS 2 interface type hash per REP-2011 RIHS standard.
+typedef struct rosidl_type_hash_s
```

Comment:
I'm having a concern that maybe type hashes should _just_ be passed around as `RIHS1_abcdef` strings and not have a struct representation like this at all. Future RIHS versions may use a different number of bytes or byte encodings for the hash values. If I encode the string directly in codegen and then just pass that on user_data, no computation is needed to be performed on them in ordinary usage,  comparison can be done as a simple `strcmp`.

Reply:
Or maybe it's okay to have RIHS1 built in to the system like this? A communicator using a different hashing standard will either have the same count of value bits (different version will still fail comparison) or it will have the wrong number of value bits (fail to parse and still fail comparisons due to 0 "unset" version)

Reply:
We discussed this, and a data structure is nice for passing around cleanly with good copy semantics. Going to leave it like this